File: apps/api/app/schemas/rooms.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator
from apps.api.app.schemas.agents import AgentRead

RoomMode = Literal["manual", "tag", "roundtable", "orchestrator"]


class RoomCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=200)
    goal: str | None = Field(default=None)
    current_mode: RoomMode = "orchestrator"

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("name must not be blank.")
        return trimmed


class RoomRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    owner_user_id: str
    name: str
    goal: str | None
    current_mode: RoomMode
    pending_mode: RoomMode | None
    created_at: datetime
    updated_at: datetime


class RoomModeUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    mode: str = Field(min_length=1, max_length=32)

    @field_validator("mode")
    @classmethod
    def validate_mode_non_blank(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("mode must not be blank.")
        return trimmed


class RoomAgentCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    agent_id: str = Field(min_length=1, max_length=64)
    position: int | None = Field(default=None, ge=1)

    @field_validator("agent_id")
    @classmethod
    def validate_non_blank_string(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("must not be blank.")
        return trimmed


class RoomAgentRead(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    room_id: str
    agent_id: str
    agent: AgentRead
    position: int
    created_at: datetime

File: apps/web/src/lib/api/rooms.ts
"use client";

import { apiFetch } from "@/lib/api/client";
import type { AgentRead } from "@/lib/api/agents";

export type RoomMode = "manual" | "roundtable" | "orchestrator";

export type RoomRead = {
  id: string;
  owner_user_id: string;
  name: string;
  goal: string | null;
  current_mode: RoomMode;
  pending_mode: RoomMode | null;
  created_at: string;
  updated_at: string;
};

export type RoomAgentRead = {
  id: string;
  room_id: string;
  agent_id: string;
  agent: AgentRead;
  position: number;
  created_at: string;
};

type RoomCreateRequest = {
  name: string;
  goal?: string | null;
  current_mode: RoomMode;
};

export function listRooms(): Promise<RoomRead[]> {
  return apiFetch<RoomRead[]>("/api/v1/rooms", { method: "GET" });
}

export function getRoom(roomId: string): Promise<RoomRead> {
  return apiFetch<RoomRead>(`/api/v1/rooms/${roomId}`, { method: "GET" });
}

export function createRoom(payload: RoomCreateRequest): Promise<RoomRead> {
  return apiFetch<RoomRead>("/api/v1/rooms", {
    method: "POST",
    body: JSON.stringify(payload)
  });
}

export function updateRoomMode(roomId: string, mode: RoomMode): Promise<RoomRead> {
  return apiFetch<RoomRead>(`/api/v1/rooms/${roomId}/mode`, {
    method: "PATCH",
    body: JSON.stringify({ mode })
  });
}

export async function deleteRoom(roomId: string): Promise<void> {
  await apiFetch<null>(`/api/v1/rooms/${roomId}`, {
    method: "DELETE"
  });
}

export function listRoomAgents(roomId: string): Promise<RoomAgentRead[]> {
  return apiFetch<RoomAgentRead[]>(`/api/v1/rooms/${roomId}/agents`, {
    method: "GET"
  });
}

export function assignRoomAgent(roomId: string, payload: { agent_id: string; position?: number }): Promise<RoomAgentRead> {
  return apiFetch<RoomAgentRead>(`/api/v1/rooms/${roomId}/agents`, {
    method: "POST",
    body: JSON.stringify(payload)
  });
}

export async function removeRoomAgent(roomId: string, agentId: string): Promise<void> {
  await apiFetch<null>(`/api/v1/rooms/${roomId}/agents/${agentId}`, {
    method: "DELETE"
  });
}
